lambda/src/new_account_iam_role.py

```diff
@@ -0,0 +1,383 @@
+    trust_policy_json,
+    assume_role_arn=None,
+    partition="aws",
+    botocore_cache_dir=BOTOCORE_CACHE_DIR,
+):  # pylint: disable=too-many-arguments
+    """Create or update a IAM role with a trusted relationship."""
+    # Validate trust policy contains properly formatted JSON.  This is
+    # not a validation against a schema, so the JSON could still be bad.
+    try:
+        json.loads(trust_policy_json)
+    except json.decoder.JSONDecodeError as exc:
+        # pylint: disable=raise-missing-from
+        raise IamRoleInvalidArgumentsError(
+            f"'trust-policy-json' contains badly formed JSON:"
+            f"\n\t{exc}\n\tJSON input:  {trust_policy_json}"
+        )
+
+    # Validate that either role arn or an AWS profile was supplied, as one
+    # of them is needed to create a AWS session.
+    if not assume_role_arn and not aws_profile:
+        raise IamRoleInvalidArgumentsError(
+            "One of 'assume-role-arn' or 'aws-profile' is required"
+        )
+
+    # Create a session using the role arn or AWS profile.
+    session = get_session(assume_role_arn, aws_profile, botocore_cache_dir)
+    iam_resource = session.resource("iam")
+    iam_client = session.client("iam")
+
+    # Create a role using the role name and assign it an assumed trust policy
+    # with the user-supplied JSON.
+    role = iam_role_create_trust(iam_resource, iam_client, role_name, trust_policy_json)
+    if not role:
+        raise IamRoleInvalidArgumentsError(f"Unable to create '{role_name}' role.")
+
+    # Attach the permission policy(s) associated with the role.
+    policy_arn = f"arn:{partition}:iam::aws:policy/{role_permission_policy}"
+    if not iam_role_create_policy(iam_client, role, role_name, policy_arn):
+        raise IamRoleInvalidArgumentsError(
+            f"Unable to attach '{policy_arn}'to {role_name}."
+        )
+    return 0
+
+
+def lambda_handler(event, context):  # pylint: disable=unused-argument
+    """Entry point if script called by AWS Lamdba."""
+    LOG.info("Received event:\n%s", json.dumps(event))
+
+    # For the CLI entrypoint (main), argparse will ensure that required
+    # arguments are provided.  The Lambda entrypoint uses environment
+    # variables to supply "arguments" and checks for required arguments
+    # are performed here.
+
+    # Optional:  Used to create assume-role-arn.
+    assume_role_name = os.environ.get("ASSUME_ROLE_NAME")
+
+    # Required:  Used for role-name, e.g., E_READONLY or E_PROVADMIN or
+    # E_PROVREADONLY.
+    role_name = os.environ.get("ROLE_NAME")
+
+    # Required:  role-permission-policy.  AWS-managed permission policy to
+    # attach to the role.
+    permission_policy = os.environ.get("PERMISSION_POLICY")
+
+    # Required:  trust-policy-json.  JSON-formatted string with trust policy.
+    trust_policy_json = os.environ.get("TRUST_POLICY_JSON")
+
+    LOG.info(
+        "Environment variables:\n\tASSUME_ROLE_NAME=%s\n\tROLE_NAME=%s"
```

Comment:
In some cases you've suggested using 'extra' for info logging and in some cases you've suggested using it for error logging.  The 'extra' feature is useful for filtering logs ... what is it that you think we'd need to filter on?  If I can understand what it is that you want might want to filter, then I can change the logs accordingly.

You can already filter on errors.  Do you want to filter on assumed role names, role name or permission policy, whether we're successful in creating role names or permission policies or not?

Reply:
All logging needs to be structured, regardless of the log-level. No log messages should be formatted for cli presentation purposes, such as with the `\n` usage here. Create a python object containing the relevant info, and output it with `extras`. This is the primary reason for using the powertools logging library.

Reply:
If you look at https://github.com/aws-samples/aws-serverless-shopping-cart/search?q=logger.info you'll see that the logger is used both in the standard way (cli presentation purposes) and for json objects.  It can be either.

However, I can create json objects similar to the way you're using 'extras' without using 'extras' and I can get rid of the newlines.  

But again, is there something special that you think you might like to filter on?  If so, then I'll use 'extras' as well.

Reply:
Hmm, my understanding was that this logger was _always_ outputting structured json. If it's not, maybe due to some missing setup, then that would explain our misunderstanding. Usage of `extras` is not the requirement. The requirement is for all log output to be structured json. `extras` is just a convenient way to inject arbitrary objects/data, along with a simple message. Of course the message part may use string formatting for some basic context. But I don't think the message should try to format any kind of complex object.

In Cloudwatch logs, with pure structured json messages, everything can be queried and filtered and related using syntax similar to SQL.

Reply:
So if you logging an error, you expect it always to be a JSON object?  Really?

Reply:
Absolutely! This is a minimum requirement for any kind of observability. I'd recommend subscribing to the honeycomb blog for starters. Here's an example: https://www.honeycomb.io/blog/simple-structured-logging-with-nlog/